File: shum/shum-server/src/main/kotlin/model/Effect.kt
package model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.*

@Suppress("PropertyName, FunctionName", "FunctionName")
@Serializable
sealed class Effect {

    val id: String
        get() = when (this) {
            is SendMessage -> _id
        }

    @Serializable
    data class SendMessage(
        val _id: String,
        val entity: Entity,
        val message: String,
        @SerialName("reply_to") val replyTo: Long? = null
    ) : Effect() {

        constructor(entity: Entity, message: String, replyTo: Long? = null) : this(
            _id = UUID.randomUUID().toString(),
            entity = entity,
            message = message,
            replyTo = replyTo
        )


        internal fun _toRemote(): RemoteEffect.SendMessage = RemoteEffect.SendMessage(
            _id = _id,
            entity = entity.toRemote(),
            message = message,
            replyTo = replyTo,
            type = "send_message"
        )
    }

    fun toRemote(): RemoteEffect = when (this) {
        is SendMessage -> _toRemote()
    }
}

File: shum/userbot/main/model/effect.py
from enum import Enum
from typing import Optional, Dict, Any

from main.model.entity import Entity


class EffectType(str, Enum):
    SEND_MESSAGE = "send_message"


class Effect:
    def __init__(self, effect_type: str, entity: Entity, message: str, reply_to: Optional[int]):
        self.type = effect_type
        self.entity = entity
        self.message = message
        self.reply_to = reply_to

    @staticmethod
    def from_dict(data: Dict[str, Any]) -> "Effect":
        return Effect(
            effect_type=data["type"],
            entity=Entity.from_dict(data["entity"]),
            message=data["message"],
            reply_to=data.get("reply_to"),
        )

    def to_dict(self) -> Dict[str, Any]:
        return {
            "type": self.type,
            "entity": self.entity.to_dict(),
            "message": self.message,
            "reply_to": self.reply_to if self.reply_to else None,
        }
